tools/src/main/java/org/neo4j/tools/dump/TransactionLogAnalyzer.java

```diff
@@ -0,0 +1,202 @@
+import static org.neo4j.kernel.impl.transaction.log.PhysicalLogFiles.getLogVersion;
+import static org.neo4j.kernel.impl.transaction.log.entry.LogEntryByteCodes.CHECK_POINT;
+import static org.neo4j.tools.util.TransactionLogUtils.openVersionedChannel;
+
+/**
+ * Merely a utility which, given a store directory or log file, reads the transaction log(s) as a stream of transactions
+ * and invokes methods on {@link Monitor}.
+ */
+public class TransactionLogAnalyzer
+{
+    /**
+     * Receiving call-backs for all kinds of different events while analyzing the stream of transactions.
+     */
+    public interface Monitor
+    {
+        /**
+         * Called when transitioning to a new log file, crossing a log version bridge. This is also called for the
+         * first log file opened.
+         *
+         * @param file {@link File} pointing to the opened log file.
+         * @param logVersion log version.
+         */
+        default void logFile( File file, long logVersion )
+        {   // no-op by default
+        }
+
+        /**
+         * A complete transaction with {@link LogEntryStart}, one or more {@link LogEntryCommand} and {@link LogEntryCommit}.
+         *
+         * @param transactionEntries the log entries making up the transaction, including start/commit entries.
+         */
+        default void transaction( LogEntry[] transactionEntries )
+        {   // no-op by default
+        }
+
+        /**
+         * {@link CheckPoint} log entry in between transactions.
+         *
+         * @param checkpoint the {@link CheckPoint} log entry.
+         * @param checkpointEntryPosition {@link LogPosition} of the checkpoint entry itself.
+         */
+        default void checkpoint( CheckPoint checkpoint, LogPosition checkpointEntryPosition )
+        {   // no-op by default
+        }
+    }
+
+    public static Monitor all( Monitor... monitors )
+    {
+        return new CombinedMonitor( monitors );
+    }
+
+    /**
+     * Analyzes transactions found in log file(s) specified by {@code storeDirOrLogFile} calling methods on the supplied
+     * {@link Monitor} for each encountered data item.
+     *
+     * @param fileSystem {@link FileSystemAbstraction} to find the files on.
+     * @param storeDirOrLogFile {@link File} pointing either to a directory containing transaction log files, or directly
+     * pointing to a single transaction log file to analyze.
+     * @param invalidLogEntryHandler {@link InvalidLogEntryHandler} to pass in to the internal {@link LogEntryReader}.
+     * @param monitor {@link Monitor} receiving call-backs for all {@link Monitor#transaction(LogEntry[]) transactions},
+     * {@link Monitor#checkpoint(CheckPoint, LogPosition) checkpoints} and {@link Monitor#logFile(File, long) log file transitions}
+     * encountered during the analysis.
+     * @throws IOException on I/O error.
+     */
+    public static void analyze( FileSystemAbstraction fileSystem, File storeDirOrLogFile,
```

Comment:
Can be package-private

Reply:
It's a publically available tool in that component, so just because your IDE tells you that it can be package private doesn't mean it should :)